include/rmm/device_scalar.hpp

```diff
@@ -36,75 +38,152 @@ class device_scalar {
    */
   explicit device_scalar(
-      T const &initial_value, cudaStream_t stream = 0,
-      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
-      : buff{sizeof(T), stream, mr} {
-    _memcpy(buff.data(), &initial_value, stream);
+      T const& initial_value,
+      cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+      : buffer{sizeof(T), stream, mr} {
+    set_value(initial_value, stream);
   }
 
   /**
-   * @brief Copies the value from device to host, synchronizes, and returns the
-   * value.
+   * @brief Copies the value from device to host, synchronizes, and returns the value.
+   * 
+   * Synchronizes `stream` after copying the data from device to host.
+   * 
+   * @note If the stream specified to this function is different from the stream specified
+   * to the constructor, then an appropriate dependency must be inserted between the streams 
+   * (e.g. using `cudaStreamWaitEvent()` or `cudaStreamSynchronize()`) before calling this function,
+   * otherwise there may be a race condition.
    *
-   * @throws `rmm::cuda_error` If the copy fails
-   * @throws `rmm::cuda_error` If synchronizing `stream` fails
+   * @throws `rmm::cuda_error` If the copy fails.
+   * @throws `rmm::cuda_error` If synchronizing `stream` fails.
    *
-   * @return T The value of the scalar after synchronizing the stream
-   * @param stream CUDA stream on which to perform the copy
+   * @return T The value of the scalar.
+   * @param stream CUDA stream on which to perform the copy and synchronize.
    */
   T value(cudaStream_t stream = 0) const {
     T host_value{};
-    _memcpy(&host_value, buff.data(), stream);
+    _memcpy(&host_value, buffer.data(), stream);
+    RMM_CUDA_TRY(cudaStreamSynchronize(stream));
     return host_value;
   }
 
   /**
-   * @brief Copies the value from host to device and synchronizes the stream.
+   * @brief Sets the value of the `device_scalar` to the given `host_value`.
+   *
+   * @note If the stream specified to this function is different from the stream specified
+   * to the constructor, then appropriate dependencies must be inserted between the streams 
+   * (e.g. using `cudaStreamWaitEvent()` or `cudaStreamSynchronize()`) before and after calling
+   * this function, otherwise there may be a race condition.
    *
+   * Does not synchronize `stream`.
+   * 
    * @throws `rmm::cuda_error` if copying `host_value` to device memory fails
    * @throws `rmm::cuda_error` if synchronizing `stream` fails
    *
    * @param host_value The host value which will be copied to device
    * @param stream CUDA stream on which to perform the copy
    */
-  void set_value(T host_value, cudaStream_t stream = 0) {
-    _memcpy(buff.data(), &host_value, stream);
+  template <typename Dummy = void>
+  auto set_value(T host_value, cudaStream_t stream = 0)
+    -> std::enable_if_t<std::is_fundamental<T>::value, Dummy>
```

Comment:
If constexpr would be awesome here so you could have one function body, but rmm is still C++14, right?

Reply:
Yep, all of RAPIDS (and CUDA for that matter).